### PR TITLE
group and filter repositories by owner

### PR DESCRIPTION
**Pull request by outofambit**
## Overview

partly addresses #6460
reboot of #6896

filter by owner with grouping by owner

## Description

- group and display github.com repositories by organization
- filter text field searches the owner field
- feature flagged for beta currently

## Release notes

Notes: [Improved] Group and filter GitHub repositories by owner name

### CODE REVIEW

**Comment by outofambit** [comment]
marking this time sensitive since i'd like this to ship to beta this week

**Comment by outofambit** [comment]
> marking this time sensitive since i'd like this to ship to beta this week

i would still _like_ to ship this to beta this week, but we don't _need_ to and have other pending work to support currently :)

File: app/src/ui/repositories-list/group-repositories.ts [hunk 1]
```diff
@@ -28,18 +45,24 @@ export function groupRepositories(
   localRepositoryStateLookup: ReadonlyMap<number, ILocalRepositoryState>
 ): ReadonlyArray<IFilterListGroup<IRepositoryListItem>> {
   const grouped = new Map<RepositoryGroupIdentifier, Repositoryish[]>()
+  const gitHubOwners: Array<string> = []
```
Comment (shiftkey) [inline app/src/ui/repositories-list/group-repositories.ts]: Could this be a `Set<string>`? We'd get "avoid duplicate entries" for free, which is helpful later on.

File: app/src/ui/repositories-list/group-repositories.ts [hunk 2]
```diff
@@ -84,9 +120,21 @@ export function groupRepositories(
   }
 
   // NB: This ordering reflects the order in the repositories sidebar.
-  addGroup('github')
-  addGroup('enterprise')
-  addGroup('other')
+  if (enableGroupRepositoriesByOwner()) {
+    const owners = gitHubOwners.reduce((acc, val) => {
```
Comment (shiftkey) [inline app/src/ui/repositories-list/group-repositories.ts]: If we use a `Set<string>` above we don't need to do this reducing to eliminate duplicate keys:

```
const owners = [...gitHubOwners.keys()]
```

Comment (outofambit) [reply app/src/ui/repositories-list/group-repositories.ts]: good call!

**Comment by tierninho** [comment]
Tested the functionality and LGTM. Could not test Enterprise repos as it does not work in dev mode.

File: app/src/ui/repositories-list/group-repositories.ts (hunk 2 above)
Comment (outofambit) [reply app/src/ui/repositories-list/group-repositories.ts]: done in 0850f18

**Comment by say25** [comment]
Looked this over from a UI perspective, not sure if it is worth re-adding `GitHub.com` and `Enterprise` groupings with Owner grouping below that?

**Comment by shiftkey** [comment]
> not sure if it is worth re-adding `GitHub.com` and `Enterprise` groupings with Owner grouping below that?

`Enterprise` grouping should not be affected by this change. I'm indifferent to whether there's value in displaying `GitHub.com` before the org groups (I don't think we can nest three levels deep in here).

**Pull request merged**